contrib/epee/include/net/levin_protocol_handler_async.h

```diff
@@ -433,8 +428,37 @@ class async_protocol_handler
           }
           break;
         }
+
         {
+          std::string temp{};
           epee::span<const uint8_t> buff_to_invoke = m_cache_in_buffer.carve((std::string::size_type)m_current_head.m_cb);
+          m_state = stream_state_head;
+
+          // abstract_tcp_server2.h manages max bandwidth for a p2p link
+          if (!(m_current_head.m_flags & (LEVIN_PACKET_REQUEST | LEVIN_PACKET_RESPONSE)))
+          {
+            // special noise/fragment command
+            static constexpr const uint32_t both_flags = (LEVIN_PACKET_BEGIN | LEVIN_PACKET_END);
+            if ((m_current_head.m_flags & both_flags) == both_flags)
+              break; // noise message, skip to next message
+
+            if (m_current_head.m_flags & LEVIN_PACKET_BEGIN)
+              m_fragment_buffer.clear();
+
+            m_fragment_buffer.append(reinterpret_cast<const char*>(buff_to_invoke.data()), buff_to_invoke.size());
+            if (!(m_current_head.m_flags & LEVIN_PACKET_END))
+              break; // skip to next message
+
+            if (m_fragment_buffer.size() < sizeof(bucket_head2))
+            {
+              MERROR(m_connection_context << "Fragmented data too small for levin header");
+              return false;
+            }
+
+            temp = std::move(m_fragment_buffer);
```

Comment:
"A valid but unspecified state". The standard never specifies behavior on moves, so it could copy the data leaving the source unchanged (small string optimization apparently might do this).

Anyway - if the wire format is "valid" then `m_fragment_buffer` is always `clear`ed when the next fragment begins (which puts the object into a specific state from the unspecified state). Or so I thought - the `size()` checks above are not valid - the fragment buffer should be empty for that query might it might not be. I will add a `clear` after this move which will usually be a NOP. Good catch.